### ISSUE TITLE
New files are listed as existing with 1 diff

### ISSUE THREAD

**Issue by nununo** [larshp/abapGit#407]
Hello,

Several new files still not existing in the target system show listed as existing with 1 diff:

![image](https://cloud.githubusercontent.com/assets/532049/19941350/b17d7db6-a127-11e6-9be7-2bc835e836d0.png)

And if I see the diff... it clearly shows that it does not exist:

![image](https://cloud.githubusercontent.com/assets/532049/19941369/c50be156-a127-11e6-989f-6afa03fed33a.png)

Cheers,
Nuno

**Comment by larshp** [comment]
issue reproduced, only happens when files are hidden,

Hiding files:
![image](https://cloud.githubusercontent.com/assets/5888506/19941914/3b83caf6-a132-11e6-9d4d-3575fe816eda.png)

Showing files:
![image](https://cloud.githubusercontent.com/assets/5888506/19941910/36a2b10a-a132-11e6-974c-557e805220cb.png)

**Comment by nununo** [comment]
Great that you managed to reproduce it.

So I'll show the files for now then. 

Actually I much prefer to hide them since in ABAP you don't really need to know the underlying files, just the objects, so thank you for the option to hide them which I only recently found :)

Cheers,
Nuno

**Comment by larshp** [comment]
:smile: 

I guess the fix for showing "new @remote" vs "1 diff" when hiding files is pretty simple, however I do not have time to look into it this week or next.

https://sessioncatalog.sapevents.com/go/agendabuilder.sessions/?l=133&sid=37856_464065&locale=en_US

**Comment by nununo** [comment]
Wow! Congratulations!
Great to know. Everyone should be aware of what you're accomplishing.

**Comment by sbcgua** [comment]
Guys, are you sure about the direction of solution ?
"New" is a diff in fact. I'd rather fix the diff page to show new files as well as changed.
What I was thinking about recently is that this "new" marker should actually be replaced with "diff" for all cases, However there should be some marker in the line showing the state of changes in local and remote. Because it can also be both remote and local change ! It can be more convenient to observe. Something like this:

![image](https://cloud.githubusercontent.com/assets/15635498/19982650/9026f01e-a20f-11e6-90aa-63108233d091.png)

Can fix whet I have time. I have currently a go-live in my project so quite busy :(

**Comment by nununo** [comment]
@sbcgua what you suggest makes a lot of sense. I fully subscribe that approach.

**Comment by larshp** [comment]
sounds good to me :smile:

**Comment by sbcgua** [comment]
Started looking into this topic. Several questions.

1) Opinion needed :) What is this "new" state ? Is it needed (valuable) at all ? I mean e.g. "local new" can be also described as "changes to commit", and "remote new" as "changes to pull". And with "changes" I mean not only new changes but any changes. Does it look logical ? Is specific "new" marker give anything useful? Asking because it affects how to design the markers - separate "new" and "changes" gives too many variants which might be less clear. 

Code questions @larshp 

2) `lcl_file_status=>status` currently returns new local objects with `MATCH = 'X'`. Is it a bug or intended ? It is not always but just for completely new objects. Looks like a bug but maybe used somewhere.

Existing obj with 1 new file
![2016-11-04_22h10_32](https://cloud.githubusercontent.com/assets/15635498/20021627/62b21bce-a2df-11e6-9468-489bee463963.png)

Completely new obj
![2016-11-04_22h10_49](https://cloud.githubusercontent.com/assets/15635498/20021640/70d689c4-a2df-11e6-8348-da53f55ebc3e.png)

3) The same method returns empty filename for new local objects. Although ! The `lt_local` contains filename, so can be copied to the output results if required. This blocks diff preview for local new files to be displayed. I wonder if empty filename is used anyhow as a trigger somewhere ? And can I adjust it (fill filenames for new local  objs)

**Comment by larshp** [comment]
1: I think it adds value to show the "new" states. It is more informative than number of diffs/changes

2: not sure :smile: just try changing it if needed, and perhaps add some unit tests

3: yeah, just try

My thoughts: if all lines for a object has NEW = 'L', then the object is "new @local", guess something similar can be done for "new @remote"

**Comment by sbcgua** [comment]
@larshp: typically, started as an easy task but not that easy in reality :)
One of issues: there is not enough information now to say if the file was changes locally or remotely. Local file can be compared to local checksum from DB, but there is no info for remote. Probably remote check sum should be saved as well. And here is a question: as I understood the local checksum is calculated just on file data, whereas git checksum includes also git blob type and file size. Why ? Was there any intention to simplify local checksums ? Could be good if unified imho. What do you think ?

**Comment by larshp** [comment]
git checksum includes the blob type, unifying is a good idea

**Comment by sbcgua** [comment]
@larshp Have a look. If OK, I'll PR - this is an independent code in fact. `update_checksums` is not actually needed, it is more for current tests so can be removed later.

**Comment by sbcgua** [comment]
Almost done, have a look (better click) :)

w files
![2016-11-08_21h28_18](https://cloud.githubusercontent.com/assets/15635498/20113801/931af13c-a5fa-11e6-8c2c-1221f0b69e60.png)
wo files
![2016-11-08_21h28_48](https://cloud.githubusercontent.com/assets/15635498/20113803/948f4f04-a5fa-11e6-8bde-b4944e8ba9b3.png)

- all diffs are shown now (new files and also in stage page)
- modified status is shown independently for local and remote (compared to last sync state)
- internal: file_status rewritten to support all of that, improved performance and hopefully structure + a UT

Still to finish one thing - push/pull should not update all local state but only actually changed - looks easy at the first glance but I have a feeling ...

**Comment by larshp** [comment]
cool, what does the red "~" mean? Is this for deletions? And will it be possible to determine if an object is deleted locally? hmmm, perhaps a feature to add later

**Comment by sbcgua** [comment]
~ is for "mixed" - a group of files that have both "added" and "modified". It has a tooltip also. But seems that it is not obvious from design =( I'll think how to make it clearer.

deletion would be a nice feature definitely :) Actually it is easy to add now (a marker). I can even do from UI perspective. But the the object deletion itself of course is a separate topic.

**Comment by sbcgua** [comment]
![image](https://cloud.githubusercontent.com/assets/15635498/20131857/cc2aa31e-a668-11e6-8688-a594b86a7c78.png)
That works with older IEs although, for old GUI with IE7 will not work

**Comment by sbcgua** [comment]
Final look (support deleted also)

![image](https://cloud.githubusercontent.com/assets/15635498/20140076/e3c3c0b8-a691-11e6-9cc5-df3cc3db334b.png)

> Still to finish one thing - push/pull should not update all local state but only actually changed - looks easy at the first glance but I have a feeling ...

This appeared to be tricky indeed. Will PR after finish this but not sure when

**Comment by larshp** [comment]
perhaps change "X" to "D"eleted or "R"emoved?

**Comment by sbcgua** [comment]
Ok, so finally the code is coming. Here is a final final look:

![2016-11-12_13h55_08](https://cloud.githubusercontent.com/assets/15635498/20237716/0ab747fa-a8e3-11e6-86f2-3686dadeb7f6.png)

I've put some change comments to PR and also to the new changelog file.
The stuff is definitely worth testing (I did how I could :). It should work fine but ... there are lots of changes. 

P.S. At the beginning of the first start it will check local checksums and propose to update them. This should be **safe unless you have both remote and local changes**. If you do - better sync repos before the abapGit update. If you did update already ... well ... the best idea is probably to copy older abapGit and sync repos with it first. Or carefully push local changes you know to remote individually and then "Update local checksums" in advanced menu. The idea is to have changes just on one side, then "Update checksums" should produce proper result.

P.P.S There might be some ghostly bug which I cannot recreate for the moment. Sometimes some files are shown as MM (both side changes) although changed just from one side. Seems like this happens because local state in DB looses one of files. This also triggers Stage and Pull showing at the same time. If you find why or how to recreate - pls report - I cannot find it for the moment and seen it just once. Maybe no bug at all, just dev state - but I pushed some changes, then merged master branch to feature branch (outside of abapGit) and then abapGit showed MM for one file (zabapgit.pro.abap) - it was just missing from DB state for some reason.

**Comment by sbcgua** [comment]
OK, mentioned bug seems to be resolved (hopefully)
Plus, added global view diff to the main page

![image](https://cloud.githubusercontent.com/assets/15635498/20240692/8142289c-a927-11e6-9565-929716abffb5.png)

Plus, added change-side marker to diff page and now if changes are remote colors are shown vice versa

![image](https://cloud.githubusercontent.com/assets/15635498/20240696/9deb755c-a927-11e6-8913-cee6667af322.png)

One issue. `lcl_tadir=>get_object_package` does not work reliably for deleted objects. `delflag` seems to be not enough. As I understand it is set when transport is released ... not sure ... I don't know this stuff. Do you know a reliable way to check if the object is deleted (even if it is not yet released) ? I tried with `DTELs` - they don't mark `delflag` =( Programs seem to do ... completely confusing.

**Comment by larshp** [comment]
Methods `lif_object~exists` are used to determine if the object exists or not,
https://github.com/larshp/abapGit/blob/master/src/zabapgit_object_dtel.prog.abap#L78

**Comment by larshp** [comment]
@nununo can you try the latest version of abapGit, it has a lot of changes regarding showing diffs, thanks

**Comment by nununo** [comment]
@larshp I am getting ready to try it. But as far as I could understand the repositories must be previously sync'd before I do the upgrade, right? I read it here:

> P.S. At the beginning of the first start it will check local checksums and propose to update them. This should be safe unless you have both remote and local changes. If you do - better sync repos before the abapGit update. If you did update already ... well ... the best idea is probably to copy older abapGit and sync repos with it first.

So, just to be on the safe side, I'm syncing everything before doing the upgrade.

I'll probably upgrade today and provide feedback.

Thank you guys for the speed at which this is evolving!

**Comment by nununo** [comment]
@larshp, now I sync'd all the code. But I'm still worried about this:

> This should be safe unless you have both remote and local changes.

Since my repositories insist in keeping small diffs... due to different kinds of flags and small details in the generated code... we can say that they are in a state where you have both remote and local changes, right?

What's the risk of upgrading now?

**Comment by larshp** [comment]
Worst case abapGit will overwrite local changes with the remote object when doing a pull. (however currently there is a bug which prevents this)

This is the same behavior as earlier abapGit versions, so I guess you have experienced this earlier?

**Comment by nununo** [comment]
Upgraded. All went smoothly. Great new features! Thanks and congratulations!

**Comment by larshp** [comment]
:tada: 

There are a few bugs in the new version, follow https://github.com/larshp/abapGit/issues I will try to fix most of them sometime this week

**Comment by larshp** [comment]
I'm closing the issue, a lot has happened since the issue was raised. 

Just open a new issue if the issue is still there

**Comment by sbcgua** [comment]
@nununo Glad that it looks useful :)

**Comment by nununo** [comment]
I'm intrigued about one thing in this new feature.

- I changed something in the source system.
- So in the target system abapGit showed: _M
- Then I manually replicated the change directly in the target system
- And now abapGit showed: MM (locally changed and remotely changed)
- But when doing a DIFF it doesn't show any differences
- Only after the PULL did the object stopped showing as having differences

Is this the expected normal behavior now? 
Not that it is a problem. It's ok and actually interesting :)

Thanks,
Nuno

**Comment by sbcgua** [comment]
Yes, that is intended behavior ... although I also had a doubt about it and maybe it should be adjusted. Here is my logic:

- There are 3 states: actual local (1), actual remote (2) and state at the moment of last sync (3).
- To identify changes abapGit compares 1 to 3 and 2 to 3. (3) is stored in DB as a hash (because otherwise the whole file content should be stored ... maybe in future #357 )
- consequently even though (1) = (2) in your case - it is not similar to (3)

IMHO the fact that files were changed is more important than that they are equal. But that's arguable - I just miss additional arguments :) Maybe there can be some special status for this case ... maybe 2 Ms but in green ... but would not it look overcomplicated ... so here I need some external opinions :)

**Comment by nununo** [comment]
I think the way it is now makes sense :)

Maybe... I'd just move those to the end of the list of modified files... since they're in sync anyway and having them at the end would allow us to focus on the ones which really are different. But this is just a detail borderlining on luxury :)

**Comment by sbcgua** [comment]
Mmm... thinking about additional state. Smth like this (~)

![image](https://cloud.githubusercontent.com/assets/15635498/20406234/2ad16260-ad15-11e6-9e45-768be0898551.png)

But really not sure - states will become ambiguous ...
@larshp what do you think ?

**Comment by larshp** [comment]
not sure, no opinion so far
